File: src/engine/features.ts
import type { TrainingRow } from './types';

export const LORE_DIMS = 15;

// #region stage:features
// 2 hour + 7 weekday + 1 holders + 3 text-shape + 15 hashed lore words = 28 = d
export const FEATURE_NAMES: string[] = [
  'hour_sin',
  'hour_cos',
  ...Array.from({ length: 7 }, (_, i) => `dow_${i}`),
  'holders_log',
  'lore_len',
  'lore_missing',
  'name_tokens',
  ...Array.from({ length: LORE_DIMS }, (_, i) => `lore_hash_${i + 1}`),
];

export type FeatureInput = Pick<TrainingRow, 'hour' | 'dow' | 'holders' | 'lore' | 'loreMissing' | 'name'>;

export function featurize(t: FeatureInput, out = new Float64Array(FEATURE_NAMES.length), o = 0): Float64Array {
  out.fill(0, o, o + FEATURE_NAMES.length);
  out[o] = Math.sin((2 * Math.PI * t.hour) / 24);
  out[o + 1] = Math.cos((2 * Math.PI * t.hour) / 24);
  out[o + 2 + t.dow] = 1;
  out[o + 9] = Math.log1p(t.holders);

  const words = t.lore.toLowerCase().match(/[a-z0-9']+/g) ?? [];
  out[o + 10] = words.length;
  out[o + 11] = t.loreMissing ? 1 : 0;
  out[o + 12] = t.name.trim().split(/\s+/).filter(Boolean).length;

  // Hashing trick: each word lands in one of 15 buckets with a ±1 sign.
  let norm = 0;
  for (const w of words) {
    const h = fnv1a(w);
    out[o + 13 + (h % LORE_DIMS)] += (h >>> 16) & 1 ? 1 : -1;
  }
  for (let i = 0; i < LORE_DIMS; i++) norm += out[o + 13 + i] ** 2;
  if (norm > 0) for (let i = 0; i < LORE_DIMS; i++) out[o + 13 + i] /= Math.sqrt(norm);
  return out;
}
// #endregion

export function fnv1a(s: string): number {
  let h = 0x811c9dc5;
  for (let i = 0; i < s.length; i++) {
    h ^= s.charCodeAt(i);
    h = Math.imul(h, 0x01000193);
  }
  return h >>> 0;
}

/** Feature columns grouped into the families shown on the dashboard. */
export const FEATURE_GROUPS: { key: string; label: string; match: (name: string) => boolean }[] = [
  { key: 'holders_log', label: 'Holders at 48h (log scale)', match: (n) => n === 'holders_log' },
  { key: 'hour_cos', label: 'Launch hour · cosine of the 24h cycle', match: (n) => n === 'hour_cos' },
  { key: 'hour_sin', label: 'Launch hour · sine of the 24h cycle', match: (n) => n === 'hour_sin' },
  { key: 'dow', label: 'Day-of-week pattern', match: (n) => n.startsWith('dow_') },
  { key: 'lore_words', label: 'Lore wording (hashed bag of words)', match: (n) => n.startsWith('lore_hash_') },
  { key: 'lore_len', label: 'Lore length', match: (n) => n === 'lore_len' },
  { key: 'lore_missing', label: 'Lore missing', match: (n) => n === 'lore_missing' },
  { key: 'name_tokens', label: 'Words in the name', match: (n) => n === 'name_tokens' },
];

File: backend/wassily/engine/features.py
"""Feature map: 2 hour + 7 weekday + 1 holders + 3 text-shape + 15 hashed lore words = 28 = d.

Nothing derived from price, volume or liquidity is ever a feature.
"""

from __future__ import annotations

import math
import re
from dataclasses import dataclass
from typing import Protocol

LORE_DIMS = 15

FEATURE_NAMES: tuple[str, ...] = (
    "hour_sin",
    "hour_cos",
    *(f"dow_{i}" for i in range(7)),
    "holders_log",
    "lore_len",
    "lore_missing",
    "name_tokens",
    *(f"lore_hash_{i + 1}" for i in range(LORE_DIMS)),
)
D = len(FEATURE_NAMES)

_LORE_WORD = re.compile(r"[a-z0-9']+")


class FeatureInput(Protocol):
    hour: int
    dow: int
    holders: float
    lore: str
    lore_missing: bool
    name: str


@dataclass(frozen=True)
class IdeaFeatures:
    """A candidate idea being scored: holders are pinned by the caller."""

    name: str
    lore: str
    hour: int
    dow: int
    holders: float
    lore_missing: bool = False


def fnv1a(s: str) -> int:
    """32-bit FNV-1a over UTF-16 code units, matching ``charCodeAt`` on the site."""
    h = 0x811C9DC5
    data = s.encode("utf-16-le")
    for i in range(0, len(data), 2):
        h ^= data[i] | (data[i + 1] << 8)
        h = (h * 0x01000193) & 0xFFFFFFFF
    return h


def featurize(t: FeatureInput) -> list[float]:
    out = [0.0] * D
    out[0] = math.sin(2 * math.pi * t.hour / 24)
    out[1] = math.cos(2 * math.pi * t.hour / 24)
    out[2 + t.dow] = 1.0
    out[9] = math.log1p(t.holders)

    words = _LORE_WORD.findall(t.lore.lower())
    out[10] = float(len(words))
    out[11] = 1.0 if t.lore_missing else 0.0
    out[12] = float(len(t.name.split()))

    # Hashing trick: each word lands in one of 15 buckets with a +/-1 sign.
    for w in words:
        h = fnv1a(w)
        out[13 + h % LORE_DIMS] += 1.0 if (h >> 16) & 1 else -1.0
    norm = math.sqrt(sum(v * v for v in out[13:]))
    if norm > 0:
        for i in range(13, D):
            out[i] /= norm
    return out


FEATURE_GROUPS: tuple[tuple[str, str], ...] = (
    ("holders_log", "Holders at 48h (log scale)"),
    ("hour_cos", "Launch hour · cosine of the 24h cycle"),
    ("hour_sin", "Launch hour · sine of the 24h cycle"),
    ("dow", "Day-of-week pattern"),
    ("lore_words", "Lore wording (hashed bag of words)"),
    ("lore_len", "Lore length"),
    ("lore_missing", "Lore missing"),
    ("name_tokens", "Words in the name"),
)


def feature_group(name: str) -> str:
    if name.startswith("dow_"):
        return "dow"
    if name.startswith("lore_hash_"):
        return "lore_words"
    return name


def grouped_importance(importance: dict[str, float]) -> dict[str, float]:
    out: dict[str, float] = {key: 0.0 for key, _ in FEATURE_GROUPS}
    for name, value in importance.items():
        out[feature_group(name)] = out.get(feature_group(name), 0.0) + value
    return out
